cpu/kinetis_common/periph/adc.c

```diff
@@ -25,18 +25,59 @@
 #include <stdio.h>
 
 #include "cpu.h"
+#include "mutex.h"
 #include "periph/adc.h"
-#include "periph_conf.h"
 
-/* guard in case that no ADC device is defined */
-#if ADC_NUMOF
+/**
+ * @brief   Maximum clock speed
+ *
+ * The ADC clock speed must be configured to be >2MHz and <12MHz in order for
+ * the ADC to work. For ADC calibration to work as intended, the ADC clock speed
+ * must further no be >4MHz, so we target a ADC clock speed > 2 and <4MHz.
+ */
+#define ADC_MAX_CLK         (4000000U)
 
-typedef struct {
-    int max_value;
-    int bits;
-} adc_config_t;
+static mutex_t locks[] = {
+    MUTEX_INIT,
+#if ADC_DEVS > 1
+    MUTEX_INIT,
+#endif
+#if ADC_DEVS > 2
+    MUTEX_INIT
+#endif
+};
+
+static inline ADC_Type *dev(adc_t line)
+{
+    return adc_config[line].dev;
+}
+
+static inline void prep(adc_t line)
+{
+    if (dev(line) == ADC0) {
+        BITBAND_REG32(SIM->SCGC6, SIM_SCGC6_ADC0_SHIFT) = 1;
+    }
+#ifdef ADC1
+    else {
+        BITBAND_REG32(SIM->SCGC3, SIM_SCGC3_ADC1_SHIFT) = 1;
+    }
+#endif
+    mutex_lock(&locks[line]);
```

Comment:
`[line]` is wrong, should be index of the `dev` device, but the dev function gives a pointer, not index.